File: src/main/java/com/kmarzecki/communicator/security/UserDetailsServiceImpl.java
package com.kmarzecki.communicator.security;


import com.kmarzecki.communicator.model.auth.RegisterDto;
import com.kmarzecki.communicator.model.auth.UserEntity;
import com.kmarzecki.communicator.repository.RoleRepository;
import com.kmarzecki.communicator.repository.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;

import static com.kmarzecki.communicator.util.CollectionUtils.asSet;
import static com.kmarzecki.communicator.util.CollectionUtils.mapList;

/**
 * User details service implementation
 */
@Service
@AllArgsConstructor
public class UserDetailsServiceImpl implements UserDetailsService  {
    private final UserRepository userRepository;
    private final RoleRepository roleRepository;

    /** Check if user exists
     * @param username user username
     * @return whether a user exists
     */
    public boolean existsByUsername(String username) {
        return userRepository.existsByUsername(username);
    }

    /**
     * Save a user
     * @param dto registration data
     */
    public void saveUser(RegisterDto dto) {
        UserEntity user = UserEntity.builder()
                .password(new BCryptPasswordEncoder().encode(dto.getPassword()))
                .username(dto.getUsername())
                .avatar_id("DEFAULT_AVATAR")
                .email(dto.getEmail())
                .roles(asSet(roleRepository.findByName("USER_ROLE")))
                .build();
        userRepository.save(user);
    }

    /**
     * Load user by username
     * @param username username
     * @return User details
     * @throws UsernameNotFoundException if user with specified username is not found
     */
    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        UserEntity user = userRepository.findByUsername(username);
        if (user != null) {
            List<GrantedAuthority> authorities = mapList(
                    role -> new SimpleGrantedAuthority(role.getName()),
                    user.getRoles()
            );
            return new User(user.getUsername(), user.getPassword(), authorities);
        } else {
            throw new UsernameNotFoundException("Not found");
        }
    }
}

File: src/main/kotlin/com/kmarzecki/trollo/security/UserDetailsServiceImpl.kt
package com.kmarzecki.trollo.security

import com.kmarzecki.trollo.model.auth.RegisterDto
import com.kmarzecki.trollo.model.auth.Role
import com.kmarzecki.trollo.model.auth.UserEntity
import com.kmarzecki.trollo.repository.RoleRepository
import com.kmarzecki.trollo.repository.UserRepository
import com.kmarzecki.trollo.util.CollectionUtils
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserDetailsServiceImpl(
        private val userRepository: UserRepository,
        private val roleRepository: RoleRepository
) : UserDetailsService {


    /** Check if user exists
     * @param username user username
     * @return whether a user exists
     */
    fun existsByUsername(username: String?): Boolean {
        return userRepository.existsByUsername(username)
    }

    /**
     * Save a user
     * @param dto registration data
     */
    fun saveUser(dto: RegisterDto) {
        val user = UserEntity(
                id = null,
                password = BCryptPasswordEncoder().encode(dto.password),
                username = dto.username,
                avatar_id = "DEFAULT_AVATAR",
                email = dto.email,
                roles = roleRepository.findByName("USER_ROLE")?.let { setOf(it) } ?: setOf()
        )
        userRepository.save(user)
    }

    /**
     * Load user by username
     * @param username username
     * @return User details
     * @throws UsernameNotFoundException if user with specified username is not found
     */
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        userRepository.findByUsername(username)
                ?.let {
                    val authorities = CollectionUtils.mapList<Role, GrantedAuthority>(
                            { role: Role -> SimpleGrantedAuthority(role.name) },
                            it.roles
                    )
                    return User(it.username, it.password, authorities)
                } ?: throw UsernameNotFoundException("Not found")
    }
}